framework/src/headers/ip/v6/icmp/neighbor_advertisement.rs

```diff
@@ -0,0 +1,265 @@
+                     When it is not set the advertisement will not
+                     update a cached link-layer address though it will
+                     update an existing Neighbor Cache entry for which
+                     no link-layer address is known.  It SHOULD NOT be
+                     set in solicited advertisements for anycast
+                     addresses and in solicited proxy advertisements.
+                     It SHOULD be set in other solicited advertisements
+                     and in unsolicited advertisements.
+
+      Reserved       29-bit unused field.  It MUST be initialized to
+                     zero by the sender and MUST be ignored by the
+                     receiver.
+
+      Target Address
+                     For solicited advertisements, the Target Address
+                     field in the Neighbor Solicitation message that
+                     prompted this advertisement.  For an unsolicited
+                     advertisement, the address whose link-layer address
+                     has changed.  The Target Address MUST NOT be a
+                     multicast address.
+
+   Possible options:
+
+      Target link-layer address
+                     The link-layer address for the target, i.e., the
+                     sender of the advertisement.  This option MUST be
+                     included on link layers that have addresses when
+                     responding to multicast solicitations.  When
+                     responding to a unicast Neighbor Solicitation this
+                     option SHOULD be included.
+
+                     The option MUST be included for multicast
+                     solicitations in order to avoid infinite Neighbor
+                     Solicitation "recursion" when the peer node does
+                     not have a cache entry to return a Neighbor
+                     Advertisements message.  When responding to unicast
+                     solicitations, the option can be omitted since the
+                     sender of the solicitation has the correct link-
+                     layer address; otherwise, it would not be able to
+                     send the unicast solicitation in the first place.
+                     However, including the link-layer address in this
+                     case adds little overhead and eliminates a
+                     potential race condition where the sender deletes
+                     the cached link-layer address prior to receiving a
+                     response to a previous solicitation.
+
+      Future versions of this protocol may define new option types.
+      Receivers MUST silently ignore any options they do not recognize
+      and continue processing the message.
+*/
+
+const ROUTER_FLAG_POS: u8 = 0;
+const SOLICITED_FLAG_POS: u8 = 1;
+const OVERRIDE_FLAG_POS: u8 = 2;
+
+#[derive(Debug)]
+#[repr(C)]
+pub struct Icmpv6NeighborAdvertisement<T>
```

Comment:
(newb)
It seems like these are message types, but this is all contained under a `headers` directory structure.  Is that right?

I would think the `ICMP` header stuff would go in `headers`, and the messages/payloads would be somewhere else?

Reply:
the other thing to add is this PR has a lot of code duplication. i'm open for ideas on how to DRY it up.

Reply:
@scrollins I am of the mindset to get things working first and refactor later (my opinion, feel free to disagree).  The reason is that once it works and we have some e2e tests, it is easier to refactor because then we can be assured that our changes are breaking (or preferably not breaking).

I tend to take this kind of approach when treading in unfamiliar waters.  The e2e tests give you a safety net.

Reply:
Qualify that, if I really know what I am doing, then build it right the first time

Reply:
i dont think we're too far from having a decent solution for this. the only outstanding issues i have are the payload length question i posted in this PR and getting the value out of an enum. i posted that in our channel. after that, we should have something working end to end.